src/clj_irods/icat.clj

```diff
@@ -0,0 +1,207 @@
+          (delay (drop (- offset cached-offset) results))
+          (delay (take limit (drop (- offset cached-offset) results)))))
+
+      ;; we have a cached listing where the number of items is less than its limit (i.e. -- it reached the end) and an offset <= our requested offset
+      (seq finished-ranges)
+      (let [[cached-limit cached-offset] (first finished-ranges)
+            results (get-in merged-listings [cached-limit cached-offset])]
+        (if (nil? limit)
+          (delay (drop (- offset cached-offset) results))
+          (delay (take limit (drop (- offset cached-offset) results)))))
+
+      ;; we have a bounded limit and offset in the request, and a cached entry that can fulfill it
+      (and (not (nil? limit)) (seq bounded-limit-offset))
+      (let [[cached-limit cached-offset] (first bounded-limit-offset)
+            results (get-in merged-listings [cached-limit cached-offset])]
+        (delay (take limit (drop (- offset cached-offset) results))))
+      
+      :else
+      nil)))
+
+;; cached offset 10 limit 10, request offset 15 limit 5 vs limit 10
+;; 10 - (15 - 10) = 5 = maximum limit we can do, i.e. must be <= request limit
+;; cached offset 0 limit 1000; request offset 100 limit 100
+;; (1000 - (100 - 0) = 900)
+
+;; listings are {limit {offset [seq]}}
+;; our offset must be <= the offset of the merged+cached listing, regardless
+;; if our limit is nil, then either the cached limit must be nil, or (cached-limit + cached-offset > actual # of items in cached listing)
+;; if our limit is specified, then (cached-limit - (our-offset - cached-offset)) >= our-limit, or cached-limit must be nil
+
+(defn flatten-cached-listings
+  "Take a set of listings as they're cached (from all-cached-listings or filtered-cached-listings) and return a simple sequence"
+  [listings]
+  (->> listings
+       vals            ; entity-type
+       (mapcat vals)   ; info-types
+       (mapcat vals)   ; sort-column
+       (mapcat vals)   ; sort-direction
+       (mapcat vals)   ; limit
+       (mapcat vals))) ; offset
+
+;; user-group-ids
+(defn- user-group-ids*
+  [irods user zone]
+  (->> [user zone ::user-group-ids]
+       (cache/cached-or-do (:cache irods) #(icat/user-group-ids user zone))))
+
+(defn user-group-ids
+  [irods user zone]
+  (->> [user zone ::user-group-ids]
+       (cache/cached-or-agent (:cache irods) #(user-group-ids* irods user zone) (:icat-pool irods))))
+
+(defn cached-user-group-ids
+  [irods user zone]
+  (->> [user zone ::user-group-ids]
+       (cache/cached-or-nil (:cache irods))))
+
+(defn maybe-user-group-ids
+  [irods user zone]
+  (delay (deref (user-group-ids irods user zone))))
+
+;; get-item
+(defn- get-item*
+  [irods user zone path & {:keys [user-group-ids]}]
+  (let [cached-ids (cached-user-group-ids irods user zone)
+        args (cond
+               (seq user-group-ids) user-group-ids
+               cached-ids           [@cached-ids]
+               :else                [user zone])]
```

Comment:
I implemented `get-item` in clj-icat-direct to take either three or four arguments. Either form takes dirname and basename, third argument is a vector of user & group IDs to use for permissions checking OR a user and zone to use to fetch the user & group IDs. Most of the time it's better if we either force this namespace's `user-group-ids` to already be cached (line 2 of this comment) or pass them in (line 1), since those options can mean the user group IDs are cached, which the user/zone option can't do.